mitodl/open-discussions
# save favorites on course home page and course search

#### Pre-Flight checklist


#### What are the relevant tickets?

closes #2091 

#### What's this PR do?

This adds the ability to favorite and un-favorite bootcamps and courses, both in the course home page (in the carousels) and on the course search page.

Get everything to work required changing a few other things. Here's what I ended up changing:

- added redux-query mutations for favoriting / unfavoriting a bootcamp and course
- added redux-query requests for getting course, bootcamp, and a list of the users favorites
- changed the pagination limit to 1000 (just an arbitrarily large number) so we can fetch all the user's favorites in one request
- added UI to the learning result card for showing the empty or full star and triggering the mutation
- refactored learning resource drawer to use redux-query request to fetch learning resource data for display
- added query for favorites to course search page so favorited items can be merged into search results, allowing the user's favorites to be displayed there

and probably a few other odds and ends. I tried wherever possible that I noticed to refactor things so that we have switch statements for the learning resource type - the hope is that this will make it easier to add programs and xpro courses in the future. I wrote a few little testing utility functions which will hopefully help with that too.

to the reviewer: sorry there is so much code `;__;`

I sort of just kept needing one more piece to get the functionality in place.

#### How should this be manually tested?

All existing course search features should work as they did before with no problems.

Other than that, if you visit the course home page or the course search you should be able to click on the star to fa(Required)vorite courses and bootcamps. When you do this it should work correctly! haha.


#### Screenshots (if appropriate)

![somefavoritesfasdfasdfjjj](https://user-images.githubusercontent.com/6207644/62237597-3f056180-b39f-11e9-9645-f65d5f282dd1.png)
![somefavorites](https://user-images.githubusercontent.com/6207644/62237598-3f056180-b39f-11e9-80ec-e21141dfde8d.png)
